src/meta/src/hummock/compaction/overlap_strategy.rs

```diff
@@ -12,20 +12,96 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use risingwave_hummock_sdk::key::user_key;
 use risingwave_hummock_sdk::key_range::KeyRange;
 use risingwave_pb::hummock::SstableInfo;
 
+pub trait OverlapInfo {
+    fn check_overlap(&self, a: &SstableInfo) -> bool;
+    fn check_multiple_overlap(&self, others: &[SstableInfo]) -> Vec<SstableInfo>;
+    fn update(&mut self, table: &SstableInfo);
+}
+
 pub trait OverlapStrategy: Send + Sync {
     fn check_overlap(&self, a: &SstableInfo, b: &SstableInfo) -> bool;
+    fn check_multiple_overlap(
+        &self,
+        tables: &[SstableInfo],
+        others: &[SstableInfo],
+    ) -> Vec<SstableInfo> {
+        let mut info = self.create_overlap_info();
+        for table in tables {
+            info.update(table);
+        }
+        info.check_multiple_overlap(others)
+    }
+
+    fn create_overlap_info(&self) -> Box<dyn OverlapInfo>;
+}
+
+#[derive(Default)]
+pub struct RangeOverlapInfo {
+    target_range: Option<KeyRange>,
+}
+
+impl OverlapInfo for RangeOverlapInfo {
+    fn check_overlap(&self, a: &SstableInfo) -> bool {
+        match self.target_range.as_ref() {
+            Some(range) => check_table_overlap(range, a),
+            None => false,
+        }
+    }
+
+    fn check_multiple_overlap(&self, others: &[SstableInfo]) -> Vec<SstableInfo> {
+        match self.target_range.as_ref() {
+            Some(key_range) => {
+                let mut tables = vec![];
+                let overlap_begin = others.partition_point(|table_status| {
+                    user_key(&table_status.key_range.as_ref().unwrap().right)
+                        < user_key(&key_range.left)
+                });
+                if overlap_begin >= others.len() {
+                    return vec![];
+                }
+                for table in &others[overlap_begin..] {
+                    if user_key(&table.key_range.as_ref().unwrap().left)
+                        > user_key(&key_range.right)
+                    {
+                        break;
+                    }
+                    tables.push(table.clone());
```

Comment:
Can we avoid the clone and return `&SstableInfo`?

Reply:
The complex of cloning table would be far smaller than calculating score. Because the overlap size would not be large.

Reply:
Another problem is that if we pass the whole `sstableinfo` object in `CompactTask`, it would make `CompactTask` too large due the buckets of consistence hash. A better idea is only storing table id in `input_ssts` of  `CompactTask`.

Reply:
> The complex of cloning table would be far smaller than calculating score. Because the overlap size would not be large.

I am not suggesting not returning a vec. I mean we can return Vec<&'a SstableInfo> instead of `Vec<SstableInfo>`

> Another problem is that if we pass the whole sstableinfo object in CompactTask, it would make CompactTask too large due the buckets of consistence hash. A better idea is only storing table id in input_ssts of CompactTask.

Will compactor need the vnode bitmap for compaction? IMO, not passing the bitmap to the compactor will not affect correctness but may affect the compaction speed because compactor only needs to include SSTs with overlapping vnodes in one split. We can split input ssts according to the vnode info to address this issue in the future.

Reply:
return `Vec<&'a SstableInfo>` seems difficult because it's a return value from a dynamic object....
`compactor` will need vnode bitmap because it needs to know how to partition data. But compactor can get vnode bitmap from it's owned `HummockVersion`, so it's not necessary to send it in `CompactTask`.